### PR TITLE
feat(models): support fused linear cross-entropy across custom models

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Adds `logits_to_keep` and `output_hidden_states` to the top-level `forward()` of 31 custom causal / conditional-generation models so they support the memory-efficient **fused linear cross-entropy** (cut cross-entropy) path. Without these hooks, `_supports_logits_to_keep()` returns `False` and the finetune recipe silently falls back to `MaskedCrossEntropy`, which materializes the full `[batch*seq, vocab]` logits — costly for large-vocab models (e.g. gpt-oss, vocab ~201k).

# Changelog

- Add `logits_to_keep` + `output_hidden_states` to `forward()` across 31 custom models (baichuan, deepseek_v3/v32/v4, ernie4_5, gemma4_moe, glm4_moe/_lite/glm_moe_dsa, gpt_oss, hy_mt2/hy_v3, kimivl/kimi_k25_vl, ling_v2, llava_onevision, mimo_v2_flash, minimax_m2, mistral3/_vlm/mistral4, nemotron_omni/_parse, qwen2_5_omni, qwen3_moe/_next/_5_moe/_vl_moe/_omni_moe, step3p5/p7), following the existing `llama` / `nemotron_v3` pattern.
- Each `forward()` now gates the `lm_head` projection on `logits_to_keep` (all positions when `0`; otherwise the last `N`; DTensor-safe; handles 2D `[T,H]` and 3D `[B,S,H]` hidden states) and returns a `ModelOutput` carrying the final hidden states when `output_hidden_states` is set (bare-tensor returns become `CausalLMOutputWithPast`).
- Default behavior (`logits_to_keep=0`, `output_hidden_states` unset) is unchanged.
- Adds a tiny-config CPU unit test per model: asserts `_supports_logits_to_keep` is `True`, `forward(logits_to_keep=1, output_hidden_states=True)` returns full-sequence hidden states + last-token logits, and the default `forward()` still returns full logits.

Out of scope:
- `gemma4_drafter` — a speculative-decoding drafter that mandatorily requires `inputs_embeds` + `shared_kv_states`; the fused-CE contract does not apply.
- Bidirectional / retriever embedding models.
- `llama`, `qwen2`, `nemotron_v3` already supported the pattern.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read the Contributor guidelines.
- [x] New tests added (tiny-config CPU test per model; all pass locally).
- [ ] Documentation — n/a (no user-facing API change; opt-in via the `linear_ce` loss).

# Additional Information

Lets `FusedLinearCrossEntropy` (`nemo_automodel/components/loss/linear_ce.py`) engage for these models instead of silently falling back to `MaskedCrossEntropy`. Verified the pattern engages cut-CE end-to-end (no fallback; peak memory drops vs `MaskedCrossEntropy`) on Nemotron-30B and gpt-oss-20b.

### CODE REVIEW

File: nemo_automodel/components/models/gpt_oss/model.py [hunk 1]
```diff
@@ -282,10 +310,28 @@ def forward(
...
+            else:
+                logits = self.lm_head(hidden[:, slice_indices, :])
+
+        if is_thd:
             logits = logits.unsqueeze(0)
-        return logits
+
+        return CausalLMOutputWithPast(
+            logits=logits,
+            hidden_states=hidden if output_hidden_states else None,
```
[the remark this answers is not in the document]
Comment (athitten) [reply nemo_automodel/components/models/gpt_oss/model.py]: @akoumpa should we fix this ?

**Review by athitten — CHANGES_REQUESTED** [review]
LGTM, thank you! One comment to address from claude review

File: nemo_automodel/components/models/gpt_oss/model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (akoumpa) [reply nemo_automodel/components/models/gpt_oss/model.py]: Yes @athitten, adding now, thank you.

**Review by pthombre — APPROVED** [review]
